lib/task.js

```diff
@@ -189,47 +198,108 @@ Task.prototype = {
     }
 
     return header + '\r\n';
-  },
+  }
 
-  logFooter: function(success, exitCode, start, finish) {
+  logFooter(success, exitCode, start, finish) {
     // Human readable success/failure thing...
-    var humanSuccess = success ?
+    let humanSuccess = success ?
       'Successful' :
       'Unsuccessful';
 
     // Yes, date subtraction yields a Number.
-    var duration = (finish - start) / 1000;
+    let duration = (finish - start) / 1000;
 
     return this.fmtLog(
       '%s task run with exit code: %d completed in %d seconds',
       humanSuccess, exitCode, duration
     );
-  },
+  }
 
-  logSchemaErrors: function(prefix, errors) {
+  logSchemaErrors(prefix, errors) {
     return this.fmtLog(
       "%s format is invalid json schema errors:\n %s",
       prefix, JSON.stringify(errors, null, 2)
     );
-  },
-
-  completeRun: function* (success) {
-    yield this.runtime.stats.timeGen(
-      'tasks.time.completed',
-      this.runtime.queue.reportCompleted(
-        this.status.taskId, this.runId, { success: success }
-      )
-    );
-  },
+  }
+
+  /**
+   * Aborts a run that is currently running or being prepared to run (pulling images,
+   * establishing states, etc).  This will optionally write an error to the stream
+   * and then write the footer and kill states.
+   *
+   * @param {String} stat - Name of the current state to be used when generating stats for killing states
+   * @param {String} error - Option error to write to the stream prior to aborting
+   */
+  async abortRun(stat, error) {
+    if (error) {
+      this.stream.write(error);
+    }
+    //
+    // Ensure that the stream has completely finished.
+    await this.stream.end(this.logFooter(
+      false, // unsuccessful task
+      -1, // negative exit code indicates infrastructure errors usually.
+      this.taskStart, // duration details...
+      new Date()
+    ));
+
+    await this.runtime.stats.timeGen(`tasks.time.states.${stat}.killed`, this.states.killed(this));
+  }
+
+  /**
+   * Resolves a run and reports the completion to the proper exchange.
+   *
+   * If a task was aborted because of a malformed payload or worker shutdown,
+   * the task will be reported as an exception.
+   *
+   * No Action will be taken for a canceled task because it has already been canceled
+   * by an external entity. Canceling the task run within the worker allows resources
+   * to be cleaned up.
+   */
+  async completeRun(success) {
```

Comment:
So I separated out the two concerns...completing a run and aborting one.  Completing a run should only be concerned with a run that hasn't actually been aborted, otherwise abortRun() should do the right thing of either reporting an exception (worker-shutdown, malformed payload) , or reporting a failure for various reasons (docker issue, etc)

Changed in 22c327c5c3a0bdaeff33a29231384aa64b058415